IntegrationTest/ExerciseEndpointsTest.cs

```diff
@@ -0,0 +1,576 @@
+        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task CreateExercise_NoAuthentication_ShouldReturn_401()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var haskellServiceSub = scope.ServiceProvider.GetService<IHaskellService>();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        var solutionRunnerResponse = new SolutionRunnerResponse { Action = ResponseCode.Pass };
+        haskellServiceSub!.SubmitSubmission(Arg.Any<SubmissionDto>()).Returns(solutionRunnerResponse);
+        exerciseRepoSub!.InsertExerciseAsync(Arg.Any<ExerciseDto>(), Arg.Any<int>()).Returns(Result.Ok());
+
+        var requestBody = CreateExerciseRequestBody();
+
+        var response = await _client.PostAsync("/v1/exercises", requestBody);
+
+        Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task GetExercisesByAuthor_ShouldReturn_GetExercisesResponseDtoList()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        var exerciseRepoResponse = new List<GetExercisesResponseDto> { new GetExercisesResponseDto(1, "Add numbers")};
+        exerciseRepoSub!.GetExercisesAsync(Arg.Any<int>()).Returns(exerciseRepoResponse);
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+
+        var response = await _client.GetAsync("/v1/exercises");
+
+        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        var responseContent = await response.Content.ReadFromJsonAsync<List<GetExercisesResponseDto>>();
+        Assert.IsType<List<GetExercisesResponseDto>>(responseContent);
+        Assert.Equal(responseContent!.First(), exerciseRepoResponse.First());
+    }
+
+    [Fact]
+    public async Task GetExercisesByAuthor_NoExercisesFound_ShouldReturn_404()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        var exerciseRepoResponse = new List<GetExercisesResponseDto> { };
+        exerciseRepoSub!.GetExercisesAsync(Arg.Any<int>()).Returns(exerciseRepoResponse);
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+
+        var response = await _client.GetAsync("/v1/exercises");
+
+        Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task GetExercisesByAuthor_NoAuthentication_ShouldReturn_401()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        var exerciseRepoResponse = new List<GetExercisesResponseDto> { new GetExercisesResponseDto(1, "Add numbers") };
```

Comment:
might not matter but if a test get caught in the middleware you can basically delete every other part of the testcase than the response and assert line